### PR TITLE
[DependencyInjection][HttpKernel] Introduce build parameters

### PR DESCRIPTION
**Pull request by HeahDude**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | yes
| Tickets       | ~
| License       | MIT
| Doc PR        | TODO

Alternative implementation of #47608.
Add a convention for parameters named with a starting dot to be available at compile-time only:
```php
$containerBuilder->setParameter('foo'); // nothing changes
$containerBuilder->setParameter('.bar'); // will not be dumped
```
Calling `$container->getParameter('.bar')` on a compiled container will throw.

### CODE REVIEW

**Comment by chalasr** [comment]
What if one gets a `ParameterBag(Interface)` injected directly? Should it throw a generic error when trying to access any param starting with a dot?

**Comment by HeahDude** [comment]
@chalasr I thought about adding some logic in the `FrozenParameterBag`.
I didn't because:
- with the current implementation, when the container is dumped, the bag is created without the build parameters already (see https://github.com/symfony/symfony/blob/6.2/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L1567)
- when injecting `ContainerBagInterface` or `ParameterBagInterface` or `Psr\Container\ContainerInterface $parameterBag` one actually gets the dumped container under the hood, so unless I'm missing something, it's covered already
- I'm not sure we want to add complexity to handle throwing when the container is compiled but not dumped, waiting for some feedback about this.
 Also it would cause problem when the dumper gets the frozen bag as it would get such exception while trying to dump parameter values.

**Comment by chalasr** [comment]
@HeahDude Makes sense, thanks. That could be done later if needed anyways

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: `ParameterNotFoundException` would be better to me

File: src/Symfony/Component/DependencyInjection/Dumper/XmlDumper.php
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Dumper/XmlDumper.php]: I'm not sure about those checks in dumpers: to me, the build-time parameters should be already removed before dumpers are called. This means we should go with an after-removing compiler pass to remove parameters.

File: src/Symfony/Component/DependencyInjection/Container.php
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Container.php]: I'd do it differently: this method should return `[]`, and ContainerBuilder should return `$this->removedParameters`, exactly as we do for removed ids. The convention itself (the leading dot) doesn't need to be known by ContainerBuilder.

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
I like that ;)

Here are some notes about the implementation.

The drawback is that parameters are going to be "public" by default, and everybody will need to add a leading dot almost all the time. On the other hand, it's not a big deal if they don't, parameters are cheap, so it's fine to me.

The next step will likely be to find a way to help the community move from public params to private ones. Here is an idea: when both 'foo' and '.foo' exist, we could consider that this means that using 'foo' at runtime is deprecated. This might help bundles adopt private params.

And another step could be to find a way to completely deprecate a parameter. An idea: params starting-and-ending with a `~` would mean that: `get('foo')` would lookup for `get('~foo~')` and trigger a deprecation. That'd work with private and public params of course.

Either for this PR, or another one, as you want, if you want :)

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php
Comment (HeahDude) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: Unfortunately this exception can use a custom message as it takes the parameter name as arg.
So it's not clear why the param is not found.
Maybe a `LogicException` from the DI component, which I've missed that it's already used in the dumped container?

File: src/Symfony/Component/DependencyInjection/Dumper/XmlDumper.php
Comment (HeahDude) [reply src/Symfony/Component/DependencyInjection/Dumper/XmlDumper.php]: It seems to be working with a compiler pass 🎉, but one test is failing though and I cannot understand why, I'll try to look again later.

**Comment by HeahDude** [comment]
>The next step will likely be to find a way to help the community move from public params to private ones. Here is an idea: when both 'foo' and '.foo' exist, we could consider that this means that using 'foo' at runtime is deprecated. This might help bundles adopt private params.

>And another step could be to find a way to completely deprecate a parameter. An idea: params starting-and-ending with a ~ would mean that: get('foo') would lookup for get('~foo~') and trigger a deprecation. That'd work with private and public params of course.

Before working on this, my intention was to work on deprecating parameters (https://symfony-devs.slack.com/archives/C8WHX21K7/p1663341791619939), so I definitely agree that this should be the next step for another PR :).

**Comment by lyrixx** [comment]
I like this idea too 👍🏼 I proposed that [8 years ago](https://github.com/symfony/symfony/issues/11881#issuecomment-56247459) 👼🏼

File: src/Symfony/Component/DependencyInjection/Dumper/XmlDumper.php
Comment (HeahDude) [reply src/Symfony/Component/DependencyInjection/Dumper/XmlDumper.php]: Fixed!

**Comment by HeahDude** [comment]
Status: needs review

File: src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php
Comment (chalasr) [inline src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: Why does it need an extra run?

Comment (HeahDude) [reply src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: Because it is not safe to just move it as it might break things, but we also need it to be done at the very last moment of compilation since parameters can still be changed after the first run.
/cc @nicolas-grekas

Comment (chalasr) [reply src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: Could you illustrate with a test case? I'm not sure to understand why removing some parameters implies to resolve placeholders again, and that would prevent (re)moving this line by mistake

Comment (HeahDude) [reply src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: Sure, this is already the case with this test https://github.com/symfony/symfony/pull/47680/files#diff-ce915f72da04411c739dd98b8350d1298450bb9258a56e6ba2d072716a15f309R76, and a parameter that could be a build parameter in the future.

Comment (chalasr) [reply src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: Got it, thank you.

**Review by chalasr — DISMISSED** [review]

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/services10.php
Comment (HeahDude) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/services10.php]: I've now removed that `getRemovedParameters()` method.
Should we dump the array (now passed as option) anyway for the consistency of this check? Or should we only check for a starting dot?

**Review by HeahDude — COMMENTED** [review]
We took the chance to meet at the last Forum PHP with @nicolas-grekas to discuss another approach on this PR.
I've pushed a commit reverting some changes and adding a new implementation from the `Kernel` which will allow to use some build parameters for the dumper and deprecate some parameters later on.

File: src/Symfony/Component/HttpKernel/Kernel.php
Comment (chalasr) [inline src/Symfony/Component/HttpKernel/Kernel.php]: Are the 2 build params used by PHPDumper really worth making Kernel as aware of the concept? 
`Kernel::buildContainer()` expecting a compiler pass as argument while it used to have no parameter at all looks quite costly to me :/ If we can't find a way to avoid the signature change, my preference goes for the previous approach

File: src/Symfony/Component/HttpKernel/Kernel.php
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Kernel.php]: let's make this a private property instead, to not have to add the argument on buildContainer?

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/Kernel.php]: another idea: move adding that pass to ContainerBuilder, and expose the result as ContainerBuilder::getRemovedParameters(), which would also return the values of removed parameters to allow dumpContainer to work?

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/Kernel.php]: Sorry, another idea, simpler one  :)
Let's just NOT try to keep the list of removed parameters. In error paths, let's use this instead:
```php
if ('.' === ($name[0] ?? '')) {
    throw new LogicException(sprintf('Parameter "%s" not found. It was probably deleted during the compilation of the container.', $name));
}
```

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/Kernel.php]: And even better: move this in ParamterNotFoundException?

File: src/Symfony/Component/HttpKernel/Kernel.php
Comment (HeahDude) [reply src/Symfony/Component/HttpKernel/Kernel.php]: Fixed.

File: src/Symfony/Component/HttpKernel/Kernel.php
Comment (HeahDude) [reply src/Symfony/Component/HttpKernel/Kernel.php]: Done.

**Comment by nicolas-grekas** [comment]
I just force-pushed on your fork some changes, see second commit.
Can you please add some tests to cover the new pass, in a dedicated test class?

File: src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: I removed this part: let it be a limitation that build-time parameters won't be resolved if they're referenced after optimization passes.

File: src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: As this removes things, shouldn't it be a removing pass instead ?

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: I get the rationale, but on the other hand, this would prevent after-removing passes from using build-time parameters. We'd better allow them there I believe.

Comment (chalasr) [reply src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: Would probably deserve to be written down as a code comment to prevent it from being moved by mistake.

Comment (stof) [reply src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: if that's the intent, it should probably be registered as a low priority instead of the default priority of 0 (otherwise, most after-removing pass not defined in the DI component will still run after it)

Comment (HeahDude) [reply src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: Done.

**Review by nicolas-grekas — APPROVED** [review]

**Comment by wouterj** [comment]
On top of my message against convention based functionality in the other PR, I just learned that a *service* starting with a `.` makes it a hidden service (which only means that `debug:container` won't list it).
Are we sure we want to have the same convention, but for other functionality for parameters and services?

**Comment by nicolas-grekas** [comment]
We're wondering about how to achieve built-time parameters [since years](https://github.com/symfony/symfony/pull/47680#issuecomment-1258324066) and this is the best and only actionable idea we've had. I like it a lot personnaly: simple and effective.

**Comment by nicolas-grekas** [comment]
Any other vote @symfony/mergers ? I'd like to merge this before #48469 to prevent merge conflicts.

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php
Comment (chalasr) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]:
```suggestion
            'inline_factories_parameter' => 'container.dumper.inline_factories', // @deprecated since Symfony 6.3
            'inline_class_loader_parameter' => 'container.dumper.inline_class_loader', // @deprecated since Symfony 6.3
```

to ease BC layers' removals

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php
Comment (chalasr) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]:
```suggestion
            trigger_deprecation('symfony/dependency-injection', '6.3', 'Option "inline_factories_parameter" passed to "%s()" is deprecated, use option "inline_factories" instead.', __METHOD__);
```

same for all notices/phpdoc

**Review by chalasr — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php
Comment (HeahDude) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: Fair enough, could we agree on `To be removed in 7.0` instead?

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php
Comment (HeahDude) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: Fixed, thanks!

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php
Comment (chalasr) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: Consistency-first I'd say 🤷‍♂️ 
Screenshot 2022-12-05 at 13 27 45

Comment (HeahDude) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: OK 😅, done, thanks!

**Review by yceruto — APPROVED** [review]
Nice addition!

**Review by chalasr — APPROVED** [review]

**Comment by chalasr** [comment]
Thank you @HeahDude.

**Pull request merged**
